### ISSUE TITLE
bugprone-sizeof-expression new false positives in pre20240924 clang-tidy, not reported by pre20240917

### ISSUE THREAD

**Issue by douzzer** [llvm/llvm-project#110551]
We're seeing a slew of new false positives of this form:
```
  CC       wolfcrypt/src/src_libwolfssl_la-dilithium.lo
/tmp/tmp.4346_11739/wolfssl_test_workdir.26775/wolfssl/wolfcrypt/src/dilithium.c:5534:21: warning: suspicious usage of 'sizeof(...)' in pointer arithmetic; this scaled value will be scaled again by the '+' operator [bugprone-sizeof-expression]
5534 |             s2 = s1 + params->s1Sz / sizeof(*s1);
|                     ^                ~~~~~~~~~~~
/tmp/tmp.4346_11739/wolfssl_test_workdir.26775/wolfssl/wolfcrypt/src/dilithium.c:5534:21: note: '+' in pointer arithmetic internally scales with 'sizeof(sword32)' == 4
```
These weren't reported by `pre20240917`.

A workaround with `pre20240924` is to parenthesize thusly:
```
            s2 = s1 + (params->s1Sz / sizeof(*s1));
```
These parentheses should not be necessary -- the `/` has higher precedence than the `+` and the incumbent code pattern is quite common.

**Comment by nicovank** [comment]
#106061, CC @whisperity FYI

**Comment by whisperity** [comment]
Hah, the parentheses are a very good idea to silence these, and to be fair, I don't remember us actually planning that some parens here and there will silence the warning!

Deciding whether something is really a false positive is always a hard thing to do. Unfortunately, the code above, at least from what is visible in the report, is still suspicious, as per [SEI-CERT ARR39-C. "Do not add or subtract a scaled integer to a pointer"](https://wiki.sei.cmu.edu/confluence/display/c/ARR39-C.+Do+not+add+or+subtract+a+scaled+integer+to+a+pointer), which the improvement landed in #106061 had as a goal to match. Now, a code or a project can always decide that they want one or more rules to not apply to them.

The underlying issue here is that the connection that `s1Sz` expresses the size of (some buffer?) in bytes is **actually** connected to `*s1` only exists in the minds of the programmer. If we want to write better code that expresses invariants and design decisions cleanly, this is a design that has more downsides than expressing _"count of `decltype(*s1)` elements"_.

**Comment by douzzer** [comment]
Thanks @whisperity for the CERT link.  Good resource.

FWIW, ARR39-C doesn't seem to anticipate or address _divisive_ scaling in this construct, and is tolerant of division in a related construct (`arr[sizeof(arr)/sizeof(arr[0])])`, typo in original).  Here, the source code under test has an internal record of an object size in bytes, and is converting that quantity by division, to allow correct use as an operand in pointer arithmetic.

This is a common code pattern and not inherently dangerous.  It would be a shame to see projects disable `bugprone-sizeof-expression` globally just to avoid noise around this.  After all, it helpfully warns on various actually-dangerous code patterns!

For our part we've just patched out the problematic code, allowing us to keep it globally active:
```
--- ./clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp.dist	2024-09-24 00:16:41.000000000 -0500
+++ ./clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp	2024-09-30 14:20:01.782122506 -0500
@@ -307,7 +307,7 @@ void SizeofExpressionCheck::registerMatc
                  offsetOfExpr()))
           .bind("sizeof-in-ptr-arithmetic-scale-expr");
   const auto PtrArithmeticIntegerScaleExpr = binaryOperator(
-      hasAnyOperatorName("*", "/"),
+      hasOperatorName("*"),
       // sizeof(...) * sizeof(...) and sizeof(...) / sizeof(...) is handled
       // by this check on another path.
       hasOperands(expr(hasType(isInteger()), unless(SizeofLikeScaleExpr)),
```

**Comment by whisperity** [comment]
The crucial difference is that with `sizeof(A) / sizeof(A[0])`, the compiler can know whether `A` is a completely sized (non-VLA) array and generate appropriate warnings if it is not (e.g., if `A` had already decayed), versus in the current code, where there is no "language-level" connection between `s1Sz` and `s1`. We went a bit beyond the _exact_ wording of the CERT rule because `N / sizeof(T)` is a "scaled integer" just like `N * sizeof(T)` would be.

However, what are your ideas about the following: I'll try creating a patch that introduces a `CheckOption:` for matching the `N / sizeof(T)` case. So there is no need for you (and projects like your project, where this pattern is prevalent) to either patch the checker or outright disable it. You'll be able to use the `.clang-tidy` file to set this option to `false` and the check will behave appropriately.

(FWIW, I'm no longer working on Clang-Tidy and Static Analysis full-time, so the making of this patch will have to wait until an auspicious evening or weekend. 😏)

**Comment by douzzer** [comment]
Oh absolutely, the ideal is a config knob to give finer control.  No hurry -- we have our workaround for now, and we'll dumpster our patch and update our config as soon as your patch merges.

Thanks!
